File: src/exercises/n0063_unique_paths_ii.rs
// A robot is located at the top-left corner of a m x n grid (marked 'Start' in the diagram below).
// The robot can only move either down or right at any point in time. The robot is trying to reach
// the bottom-right corner of the grid (marked 'Finish' in the diagram below).
// Now consider if some obstacles are added to the grids. How many unique paths would there be?

/// ## Example 1:
//
//  Input:
//  [
//    [0,0,0],
//    [0,1,0],
//    [0,0,0]
//  ]
//  Output: 2
//  Explanation:
//  There is one obstacle in the middle of the 3x3 grid above.
//  There are two ways to reach the bottom-right corner:
//  1. Right -> Right -> Down -> Down
//  2. Down -> Down -> Right -> Right

#[allow(dead_code)]
static NEED_COMPILE: bool = false;

#[allow(dead_code)]
pub struct Solution {}

impl Solution {
    #[allow(dead_code)]
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        let m = obstacle_grid.len();
        let n = obstacle_grid[0].len();
        let mut grids = vec![vec![0; n + 1]; m + 1];
        grids[1][0] = 1;
        for i in 1..=m {
            for j in 1..=n {
                if obstacle_grid[i - 1][j - 1] != 1 {
                    grids[i][j] = grids[i][j - 1] + grids[i - 1][j];
                }
            }
        }
        grids[m][n]
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    pub fn test_62() {
        assert_eq!(
            Solution::unique_paths_with_obstacles(vec![
                vec![0, 0, 0],
                vec![1, 1, 0],
                vec![0, 0, 0]
            ]),
            1
        );
        assert_eq!(
            Solution::unique_paths_with_obstacles(vec![
                vec![0, 0, 0],
                vec![0, 1, 0],
                vec![0, 0, 0]
            ]),
            2
        );
        assert_eq!(Solution::unique_paths_with_obstacles(vec![vec![1]]), 0);
        assert_eq!(
            Solution::unique_paths_with_obstacles(vec![vec![0, 1], vec![1, 0]]),
            0
        );
    }
}

File: src/n0063_unique_paths_ii.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
    if len(obstacleGrid) == 0 || len(obstacleGrid[0]) == 0 {
        return 0
    }
    m, n := len(obstacleGrid), len(obstacleGrid[0])
    dp := make([][]int, m+1)
    for i := range dp {
        dp[i] = make([]int, n+1)
    }

    for i := m-1; i >= 0; i -= 1 {
        for j := n-1; j >= 0; j -= 1 {
            if i == m-1 && j == n-1 {
                if obstacleGrid[i][j] == 1 {
                    return 0
                }
                dp[i][j] = 1
                continue
            }
            if obstacleGrid[i][j] == 1 {
                continue
            }
            dp[i][j] = dp[i+1][j] + dp[i][j+1]
        }
    }
    return dp[0][0]
}
